### ISSUE TITLE
Action not picking up latest build image

### ISSUE THREAD

**Issue by freshjwarren** [bcgov/moodle#44]

**Comment by kkapoor-fresh** [comment]
- Investigated and troubleshooted the code.
- Checked if triggers are correctly configured
- Checked if the image is correctly uploaded to the image stream

**Comment by kkapoor-fresh** [comment]
I made some changes to fix. I did some testing at my end to validate that as well. However, I think it would be better if we can test by making changes to the moodle app and see if the changes get reflected after the pipeline completes successfully. Will work with Warren to make the changes and test the pipeline run

**Comment by warrenchristian1telus** [comment]
I've added the changes and successfully run the pipeline. I've added an ImageChange trigger on the Build Config, but it didn’t seem to be working, and the ConfigChange didn’t seem to trigger a rebuild either. Not quite sure what’s missing or misconfigured yet, but I was able to successfully refresh the environment by deleting the pod.

The Adaptable theme / plugin has been installed as expected.

However, I'm not sure why the Moodle version is not latest. Latest build is 2021-22-03, but we are still on build 2021-09-13, even though it was built today.

![image](https://user-images.githubusercontent.com/70002196/135925378-5e512bfc-5409-4638-aaae-77d139e36241.png)

@kkapoor-fresh I think it may be cached according to this log message: "imagestream.image.openshift.io/moodle-gamma unchanged"

https://github.com/bcgov/moodle/runs/3795036624?check_suite_focus=true#step:5:24

I've added "--no-cache" to the start-build command, in an attempt to get the latest updates for Moodle.

https://github.com/bcgov/moodle/blob/8f2387a94d8501315a210756d5a3601617d04cdb/.github/workflows/deploy.yml#L66

**Comment by kkapoor-fresh** [comment]
@warrenchristian1telus 
The image change trigger is not required in this case because we do not automatically want to trigger the Deployment config once the build step in the pipeline is complete. The Deploy to Dev step in the GA pipeline is responsible for fetching the latest image from the image stream - "moodle-gamma" defined in the Deployment config template. So, when we run the template for deployment config, it should configure a new deploymentconfig and deploy the app.

The message  "imagestream.image.openshift.io/moodle-gamma unchanged"  is correct as the image stream template inside the buildconfig resource defined in openshift/spp/docker-build doesn't change. The thing that is going to change is the image inside the imagestream tag .

Regarding the --no-cache, as per my understanding, I am creating a new build config resource everytime the pipeline runs. When this happens, the subsequent star-build command should be building the moodle image from scratch.

**Comment by warrenchristian1telus** [comment]
Well, something has been keeping updates from reaching the pod. Any suggestions as to what else that might be? I can’t see how the pod will get updated if it doesn’t get replaced when the content changes in a build. It seemed to be building the image, but the pod never gets replaced.

Also, not sure if it’s another issue, but I saved a change and the build/action didn’t run. I tried re-running the last job, and it seems to be stuck in the queue:

***@***.***

Warren Christian
Senior Developer / Analyst, TELUS Health & Payment Solutions
T 1 250 686 1359
telus.com<https://telus.com/>  Facebook<https://www.facebook.com/telus>  Twitter<https://twitter.com/telus>  Instagram<https://www.instagram.com/telus/>  YouTube<https://www.youtube.com/telus>  LinkedIn<https://ca.linkedin.com/company/telus>

This email, including any attachments, is for the sole use of the intended recipient and may contain confidential information. If you are not the intended recipient, please notify us immediately and destroy this email and any copies.

**Comment by kkapoor-fresh** [comment]
I have observed that sometimes the GITHUB ACTIONS Jobs get queued up due to issues with GA on Github side.

On your other question, yes there is something that is preventing the pods from being replaced and recreated. Ideally, when the deployment config resource is created it shouls be recreating the replication controller and pods. I am thinking to do an explicit rollout of the deployment config using oc rollout command in the Actions Deplou to Dev step

**Comment by kkapoor-fresh** [comment]
@warrenchristian1telus There was a syntactical error while deploying the build when image stream changes were added to the template file. Did you manage to fix it already?
![image](https://user-images.githubusercontent.com/77285719/136057365-d194ef8c-a0fd-47a0-b791-683d56f2a3da.png)

**Comment by warrenchristian1telus** [comment]
@kkapoor-fresh I believe I fixed the syntax issue. It's re-building now, so we should find out soon if it worked.

**Comment by kkapoor-fresh** [comment]
The Deployment config is now picking up the latest build via Github actions.
Here's the image created by the build
![image](https://user-images.githubusercontent.com/77285719/136089395-bc68ae18-4031-4993-a3d4-172ebdc5b21e.png)

Here's the deployment config/pod referencing the latest build image
![image](https://user-images.githubusercontent.com/77285719/136089526-ac0d8653-b8a6-4bda-9352-cc60a00d97ff.png)

**Comment by warrenchristian1telus** [comment]
Looks good @kkapoor-fresh.  I'm still getting the newer version alert though:

![image](https://user-images.githubusercontent.com/70002196/136100172-d9ed45ff-6954-4fc5-b21b-57f46c61bb48.png)

However, on the Releases for Moodle 3.9, it looks like we have the latest:

![image](https://user-images.githubusercontent.com/70002196/136100289-5a080178-5d73-4192-8fd8-6e4fbb7f12fb.png)

The only discrepancy I see is that the newer version seems to have a ".03" at the end: 2020061510.03

And I do see the updates three days ago in the git logs:

https://git.moodle.org/gw?p=moodle.git;a=log;h=refs/heads/MOODLE_39_STABLE;opt=--no-merges

I've tried adjusting the moodle version requirement in composer.json to see if I can get it to pick up the latest updates. It's pretty close, so shouldn't be too big of a deal, but we'll need to be able to get the latest security patches installed at some point.
